tests/zfs-tests/tests/functional/tmpfile/tmpfile_test.c

```diff
@@ -19,28 +19,26 @@ int
 main(int argc, char *argv[])
 {
 	int fd;
-	struct stat buf;
 
 	if (argc < 2) {
-		fprintf(stderr, "Usage: %s dir\n", argv[0]);
-		return (2);
-	}
-	if (stat(argv[1], &buf) < 0) {
-		perror("stat");
-		return (2);
-	}
-	if (!S_ISDIR(buf.st_mode)) {
-		fprintf(stderr, "\"%s\" is not a directory\n", argv[1]);
+		fprintf(stderr, "Usage: %s file\n", argv[0]);
 		return (2);
 	}
 
 	fd = open(argv[1], O_TMPFILE | O_WRONLY, 0666);
 	if (fd < 0) {
 		/*
-		 * Only fail on EISDIR. If we get EOPNOTSUPP, that means
-		 * kernel support O_TMPFILE, but the path at argv[1] doesn't.
+		 * Expected return values from open(2):
+		 *
+		 * ENOENT: pathname refers to a nonexistent directory,
+		 * O_TMPFILE and one of O_WRONLY or O_RDWR were specified
+		 * in flags, but this kernel version does not provide the
+		 * O_TMPFILE * functionality.
+		 *
+		 * EOPNOTSUPP: The filesystem containing pathname does not
+		 * support O_TMPFILE.
 		 */
-		if (errno == EISDIR) {
+		if (errno == ENOENT || errno == EOPNOTSUPP) {
```

Comment:
http://man7.org/linux/man-pages/man2/open.2.html
> One must check for two different error codes, EISDIR and ENOENT, when trying to determine whether the kernel supports O_TMPFILE functionality.

I don't know why I didn't check for both EISDIR and ENOENT, but we shouldn't check for EOPNOTSUPP because that flag means kernel supports it, but filesystem doesn't. And as long as kernel supports, we should not skip the test.

Reply:
Interesting, I should have dug a little deeper before assuming the new CentOS 7.5 kernel didn't support it.  Apparently it does, in a non-standard way, to preserve the existing KABI.  We could probably update ZFS to use the extended interface with a little work.  But it would probably be better for compatibility to add a way to make it clear we know that the kernel interface is available but we've intentionally decided to not use it.

@tuxoko would you mind taking over this issue and sorting it out?  I'll just close this PR.

Reply:
@behlendorf 
Can you point out where I can find about what CentOS does differently than others?

Reply:
You can download the full kernel source here.  

https://wiki.centos.org/HowTos/I_need_the_Kernel_Source#head-a8dae925eec15786df9f6f8c918eff16bf67be0d